network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -114,367 +114,276 @@ func NewControlMsgValidationInspector(
-// the specific issue encountered.
-// The returned error can be of two types:
-//  1. Expected errors: These are issues that are expected to occur during normal
-//     operation, such as invalid messages or messages that don't follow the
-//     conventions. These errors should be handled gracefully by the caller.
-//  2. Exceptions: These are unexpected issues, such as internal system errors
-//     or misconfigurations, that may require immediate attention or a change in
-//     the system's behavior. The caller should log and handle these errors
-//     accordingly.
-//
-// The returned error is returned to the gossipsub node which causes the rejection of rpc (for non-nil errors).
+// Inspect is called by gossipsub upon reception of a rpc from a remote  node.
+// It creates a new InspectRPCRequest for the RPC to be inspected async by the worker pool.
 func (c *ControlMsgValidationInspector) Inspect(from peer.ID, rpc *pubsub.RPC) error {
-	control := rpc.GetControl()
-	for _, ctrlMsgType := range p2pmsg.ControlMessageTypes() {
-		lg := c.logger.With().
+	// queue further async inspection
+	req, err := NewInspectRPCRequest(from, rpc)
+	if err != nil {
+		c.logger.Error().
+			Err(err).
 			Str("peer_id", from.String()).
-			Str("ctrl_msg_type", string(ctrlMsgType)).Logger()
-		validationConfig, ok := c.config.GetCtrlMsgValidationConfig(ctrlMsgType)
-		if !ok {
-			lg.Trace().Msg("validation configuration for control type does not exists skipping")
-			continue
-		}
-
-		switch ctrlMsgType {
-		case p2pmsg.CtrlMsgGraft, p2pmsg.CtrlMsgPrune:
-			// normal pre-processing
-			err := c.blockingPreprocessingRpc(from, validationConfig, control)
-			if err != nil {
-				lg.Error().
-					Err(err).
-					Msg("could not pre-process rpc, aborting")
-				return fmt.Errorf("could not pre-process rpc, aborting: %w", err)
-			}
-		case p2pmsg.CtrlMsgIHave:
-			// iHave specific pre-processing
-			sampleSize := util.SampleN(len(control.GetIhave()), c.config.IHaveInspectionMaxSampleSize, c.config.IHaveSyncInspectSampleSizePercentage)
-			err := c.blockingIHaveSamplePreprocessing(from, validationConfig, control, sampleSize)
-			if err != nil {
-				lg.Error().
-					Err(err).
-					Msg("could not pre-process rpc, aborting")
-				return fmt.Errorf("could not pre-process rpc, aborting: %w", err)
-			}
-		}
-
-		// queue further async inspection
-		req, err := NewInspectMsgRequest(from, validationConfig, control)
-		if err != nil {
-			lg.Error().
-				Err(err).
-				Str("peer_id", from.String()).
-				Str("ctrl_msg_type", string(ctrlMsgType)).
-				Msg("failed to get inspect message request")
-			return fmt.Errorf("failed to get inspect message request: %w", err)
-		}
-		c.workerPool.Submit(req)
+			Msg("failed to get inspect RPC request")
+		return fmt.Errorf("failed to get inspect RPC request: %w", err)
 	}
-
+	c.workerPool.Submit(req)
 	return nil
 }
 
-// Name returns the name of the rpc inspector.
-func (c *ControlMsgValidationInspector) Name() string {
```

Comment:
While the `ControlMsgValidationInspector` is covered by integration tests in the `insecure` package, its critical role in our codebase warrants dedicated unit tests. These tests should focus on both successful and unsuccessful RPC processing:

1. Verify that legitimate RPCs (iHave, iWant, Graft, Prune) do not trigger misbehavior notifications.
2. Test for iHave misbehaviors, ensuring each case triggers a misbehavior notification.
3. Similarly, validate that iWant misbehaviors trigger misbehavior notifications.
4. Confirm that GRAFT and PRUNE misbehaviors also lead to misbehavior notifications.

Please include these tests in this PR or create an issue for them to be implemented in a subsequent PR.

Reply:
Currently in progress here https://github.com/dapperlabs/flow-internal/issues/1902